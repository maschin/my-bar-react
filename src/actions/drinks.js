import demoDrinks from '../tests/fixtures/drinks';

//ADD DRINK
export const addDrink = (drink) => ({
    type: 'ADD_DRINK',
    drink
});

export const startAddDrink = (drink = {}) => {
    return (dispatch, getState) => {
        drink.id = '123'+Date.now();
        dispatch(addDrink(drink));
    }
};

//EDIT DRINK
export const editDrink = (id, updates) => ({
    type: 'EDIT_DRINK',
    id,
    updates
});

export const startEditDrink = (id, updates) => {
    return (dispatch, getState) => {
        dispatch(editDrink(id, updates));
    }
};

//REMOVE DRINK
export const removeDrink = ( { id } = {}) => ({
    type: 'REMOVE_DRINK',
    id
});

export const startRemoveDrink = ( { id } = {}) => {
    return (dispatch, getState) => {
        dispatch(removeDrink({ id }));
    }
};

//SET DRINKS
export const setDrinks = (drinks) => ({
    type: 'SET_DRINKS',
    drinks
});

export const startSetDrinks = () => {
    return (dispatch, getState) => {
        dispatch(setDrinks(demoDrinks));
    }
}