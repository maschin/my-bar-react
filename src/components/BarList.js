import React from 'react';
import { connect } from 'react-redux';

import DrinkListItem from '../components/DrinkListItem';

export const BarList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-desktop">Drink</div>
            <div className="show-for-desktop">Data</div>
        </div>
        <div className="list-body">
            {
                props.drinks.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>Your bar is totally empty!</span>
                    </div>
                ) : (
                    props.drinks.map((drink) => (
                       <DrinkListItem key={drink.id} {...drink} />
                    ))
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks
    };
};

export default connect(mapStateToProps)(BarList);