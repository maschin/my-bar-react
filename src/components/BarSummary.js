import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const BarSummary = ({ bottlesCount }) => {
    const bottlesWord = bottlesCount === 1 ? 'bottle' : 'bottles';

    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    You have total <span>{bottlesCount}</span> {bottlesWord} in your bar.
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Drink</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks,
        bottlesCount: state.drinks.length
    };
};

export default connect(mapStateToProps)(BarSummary);