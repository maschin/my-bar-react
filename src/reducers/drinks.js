const drinksReducerDefaultState = [
    {
        id: 'sdfsdf',
        title: 'My bottle',
        category: 'Beer',
        vol: 0.5,
        alc: 5.6,
        price: 450,
        aged: 12,
        boughtAt: 123123,
        bottledAt: 12312312,
        bestBefore: 12312312,

    },
    {
        id: 'sdfsdf',
        title: 'My bottle',
        category: 'Beer',
        vol: 0.5,
        alc: 5.6,
        price: 450,
        aged: 12,
        boughtAt: 123123,
        bottledAt: 12312312,
        bestBefore: 12312312,

    }
];

export default (state = drinksReducerDefaultState, action) => {
    switch (action.type){
        case 'SET_DRINKS':
            return action.drinks;
        default:
    }       return state;
}