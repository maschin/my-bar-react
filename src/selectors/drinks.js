//Get visible drinks

export default (drinks, { text, sortBy }) => {
    return drinks.filter((drink) => {
        const textMatch = drink.title.trim().toLowerCase().includes(text.trim().toLowerCase()) || drink.category.trim().toLowerCase().includes(text.trim().toLowerCase());

        return textMatch;
    }).sort((a, b) => {
        switch(sortBy){
            default:
                return a.createdAt < b.createdAt ? 1 : -1;
        }
    })
}
