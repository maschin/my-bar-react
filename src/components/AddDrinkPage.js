import React from 'react';
import { connect } from 'react-redux';

import DrinkForm from '../components/DrinkForm';
import { startAddDrink } from '../actions/drinks';

export class AddDrinkPage extends React.Component {
    onSubmit = (drink) => {
        this.props.startAddDrink(drink);
        this.props.history.push('/');
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Bottle</h1>
                    </div>
                </div>
                <div className="content-container">
                    <DrinkForm
                        actionTitle='Add Bottle'
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddDrink: (drink) => dispatch(startAddDrink(drink))
});

export default connect(undefined, mapDispatchToProps)(AddDrinkPage);