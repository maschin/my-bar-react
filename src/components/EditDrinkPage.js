import React from 'react';
import { connect } from 'react-redux';
import DrinkForm from '../components/DrinkForm';
import {startEditDrink, startRemoveDrink} from '../actions/drinks';

export class EditDrinkPage extends React.Component {
    onRemove = () => {
        this.props.startRemoveDrink(this.props.drink);
        this.props.history.push('/');
    }
    onSubmit = (drink) => {
        this.props.startEditDrink(this.props.drink.id, drink);
        this.props.history.push(`/view/${this.props.drink.id}`);
    }
    render(){
        return (
            <div>
                {
                    this.props.drink ? (
                        <div>
                            <div className="page-header">
                                <div className="content-container">
                                    <h1 className="page-header__title">Edit Drink</h1>
                                    <button id="remove" className="button" onClick={this.onRemove}>Delete</button>
                                </div>
                            </div>
                            <div className="content-container">
                                <DrinkForm
                                    actionTitle='Save'
                                    onSubmit={this.onSubmit}
                                    drink={this.props.drink}
                                />
                            </div>
                        </div>
                    ) : (
                        <div className="content-container">
                            <h1 className="page-header__title">Record with <span>id:{this.props.match.params.id}</span> not found!</h1>
                        </div>
                    )}
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startEditDrink: (id, updates) => dispatch(startEditDrink(id, updates)),
    startRemoveDrink: (data) => dispatch(startRemoveDrink(data))
});

const mapStateToProps = (state, props) => ({
    drink: state.drinks.find((drink) => (drink.id === props.match.params.id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditDrinkPage);