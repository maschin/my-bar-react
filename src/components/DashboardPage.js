import React from 'react';
import BarList from '../components/BarList';
import BarSummary from '../components/BarSummary';

export const DashboardPage = () => (
    <div>
        <BarSummary />
        <BarList />
    </div>
);

export default DashboardPage;