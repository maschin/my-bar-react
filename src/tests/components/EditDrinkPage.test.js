import React from 'react';
import { shallow } from 'enzyme';
import { EditDrinkPage } from '../../components/EditDrinkPage';
import drinks from '../fixtures/drinks';

let wrapper, history, startEditDrink;

beforeEach(() => {
    startEditDrink = jest.fn();
    history = { push: jest.fn() }
    wrapper = shallow(
        <EditDrinkPage
            drink={drinks[1]}
            startEditDrink={startEditDrink}
            history={history}
        />
    );
});

test('should render edit page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render edit page with invalid id correctly', () => {
    const params = { id: 123 };
    const match = { params };
    wrapper = shallow(<EditDrinkPage match={ match } />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle on submit', () => {
    wrapper.find('DrinkForm').prop('onSubmit')(drinks[1]);
    expect(startEditDrink).toHaveBeenCalledWith(drinks[1].id, drinks[1]);
    expect(history.push).toHaveBeenCalledWith(`/view/${drinks[1].id}`);
});

test('should handle on delete', () => {
    const startRemoveDrink = jest.fn();
    const history = { push: jest.fn()};
    const wrapper = shallow(<EditDrinkPage drink={drinks[1]} startRemoveDrink={startRemoveDrink} history={history} />);
    wrapper.find('#remove').simulate('click');
    expect(history.push).toHaveBeenCalledWith('/');
    expect(startRemoveDrink).toHaveBeenCalledWith(drinks[1]);
});
