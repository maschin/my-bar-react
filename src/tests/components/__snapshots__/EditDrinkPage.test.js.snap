// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit page correctly 1`] = `
<div>
  <div>
    <div
      className="page-header"
    >
      <div
        className="content-container"
      >
        <h1
          className="page-header__title"
        >
          Edit Drink
        </h1>
      </div>
    </div>
    <div
      className="content-container"
    >
      <DrinkForm
        actionTitle="Save"
        categories={
          Array [
            "Absinthe",
            "Beer",
            "Brandy",
            "Cider",
            "Gin",
            "Horilka",
            "Liqueurs",
            "Mastika",
            "Ouzo",
            "Rakia",
            "Slivovitz",
            "Rum",
            "Sparkling Wine",
            "Tequila",
            "Vodka",
            "Vermouth",
            "Wine",
            "Whisky",
          ]
        }
        drink={
          Object {
            "aged": 12,
            "alc": 42,
            "bestBefore": 10368000000,
            "bottledAt": -126230400000,
            "boughtAt": -31536000000,
            "category": "Whisky",
            "createdAt": 0,
            "description": "",
            "id": "whisky",
            "isEmpty": false,
            "isOpen": true,
            "price": "",
            "title": "My bottle II",
            "updatedAt": 0,
            "vol": 0.7,
          }
        }
        onSubmit={[Function]}
        options={
          Object {
            "validate": true,
          }
        }
        vol={
          Array [
            0.25,
            0.33,
            0.375,
            0.5,
            0.6,
            0.7,
            0.75,
            1,
            5,
          ]
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render edit page with invalid id correctly 1`] = `
<div>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Record with 
      <span>
        id:
        123
      </span>
       not found!
    </h1>
  </div>
</div>
`;
